packages/instantsearch-ui-components/src/components/Carousel.tsx

```diff
@@ -0,0 +1,222 @@
+/** @jsx createElement */
+import { cx } from '../lib';
+
+import type {
+  ComponentProps,
+  RecommendItemComponentProps,
+  RecordWithObjectID,
+  Renderer,
+} from '../types';
+
+export type CarouselProps<
+  TObject,
+  TComponentProps extends Record<string, unknown> = Record<string, unknown>
+> = ComponentProps<'div'> & {
+  items: Array<RecordWithObjectID<TObject>>;
+  itemComponent: (
+    props: RecommendItemComponentProps<RecordWithObjectID<TObject>> &
+      TComponentProps
+  ) => JSX.Element;
+  classNames?: Partial<CarouselClassNames>;
+  translations?: Partial<CarouselTranslations>;
+};
+
+export type CarouselClassNames = {
+  /**
+   * Class names to apply to the root element
+   */
+  root: string | string[];
+  /**
+   * Class names to apply to the list element
+   */
+  list: string | string[];
+  /**
+   * Class names to apply to each item element
+   */
+  item: string | string[];
+  /**
+   * Class names to apply to both navigation elements
+   */
+  navigation: string | string[];
+  /**
+   * Class names to apply to the next navigation element
+   */
+  navigationNext: string | string[];
+  /**
+   * Class names to apply to the previous navigation element
+   */
+  navigationPrevious: string | string[];
+};
+
+export type CarouselTranslations = {
+  /**
+   * The label of the next navigation element
+   */
+  nextButtonLabel: string;
+  /**
+   * The title of the next navigation element
+   */
+  nextButtonTitle: string;
+  /**
+   * The label of the previous navigation element
+   */
+  previousButtonLabel: string;
+  /**
+   * The title of the previous navigation element
+   */
+  previousButtonTitle: string;
+  /**
+   * The label of the carousel
+   */
+  listLabel: string;
+};
+
+let lastCarouselId = 0;
+
+function generateCarouselId() {
+  return `ais-Carousel-${lastCarouselId++}`;
+}
+
+export function createCarouselComponent({ createElement, useRef }: Renderer) {
+  // This prevents the SVG buttons references from being hoisted
+  // during the build, where `createElement` is not available.
+  const buttonPathCommonProps: JSX.IntrinsicElements['path'] = {
+    fillRule: 'evenodd',
+    clipRule: 'evenodd',
+    fill: 'currentColor',
+  };
+
+  return function Carousel<TObject extends RecordWithObjectID>(
+    props: CarouselProps<TObject>
+  ) {
+    const listRef = useRef?.<HTMLOListElement>(null);
+    const nextButtonRef = useRef?.<HTMLButtonElement>(null);
+    const previousButtonRef = useRef?.<HTMLButtonElement>(null);
+    const carouselIdRef = useRef?.(generateCarouselId());
```

Comment:
📝 this will need to be addressed when we'll use it for Vue

Reply:
We should probably just pass all the refs, I don't see how passing `useRef` to be called there is necessary
But yeah this could be changed later

Reply:
+1. This leaks the `useRef` Hook API in the generic UI component. We could accept `listRef`, `nextButtonRef`, `previousButtonRef` and `carouselIdRef` as props. Changing this only when Vue is supported might be a breaking change?

Reply:
I think both options are not agnostic, as they assume the reference adheres to a specific structure (`{ current: T }`) in order to be bound to a DOM element.

Instead, passing refs as props add complexity on the developers' end if we decide to provide the Carousel as a wrapper (we can handle the complexity if we use it internally on our widgets).

Proposed changes are done in [1d51ef4](https://github.com/algolia/instantsearch/pull/6289/commits/1d51ef47c479c46b4cf42a9827e3e990cb119fe0).

Reply:
My main issue was that in Vue we call the `create` function each render, and calling `useRef` also makes these components not stateless anymore.

Changes look good to me, should make supporting Vue easier later I believe :)